src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/roundabout/ComplexRoundabout.java

```diff
@@ -0,0 +1,306 @@
+            try
+            {
+                this.roundaboutRoute = Route.fromNonArrangedEdgeSet(this.roundaboutEdgeSet, false);
+            }
+            // If a Route cannot be formed, invalidate the roundabout.
+            catch (final CoreException badRoundabout)
+            {
+                this.invalidationReasons.add(INCOMPLETE_ROUTE_INVALIDATION);
+                throw new CoreException(
+                        "Exception thrown building Route for Edges {}: {}", this.roundaboutEdgeSet
+                                .stream().map(Edge::getIdentifier).collect(Collectors.toSet()),
+                        INCOMPLETE_ROUTE_INVALIDATION);
+            }
+
+            // Invalidate the roundabout if any of the edges are not car navigable or master Edges,
+            // or the route does not
+            // form a closed loop.
+            if (this.roundaboutEdgeSet.stream()
+                    .anyMatch(roundaboutEdge -> !HighwayTag.isCarNavigableHighway(roundaboutEdge)
+                            || !roundaboutEdge.isMasterEdge())
+                    || !this.roundaboutRoute.start().inEdges().contains(this.roundaboutRoute.end()))
+            {
+                this.invalidationReasons.add(INCOMPLETE_ROUTE_INVALIDATION);
+            }
+
+            // Get the direction of the roundabout
+            final RoundaboutDirection direction = findRoundaboutDirection(this.roundaboutRoute);
+
+            // Determine if the roundabout is in a left or right driving country
+            final boolean isLeftDriving = leftDrivingCountries.contains(isoCountryCode);
+
+            // If the roundabout traffic is clockwise in a right-driving country, or
+            // If the roundabout traffic is counterclockwise in a left-driving country
+            if (direction.equals(RoundaboutDirection.CLOCKWISE) && !isLeftDriving
+                    || direction.equals(RoundaboutDirection.COUNTERCLOCKWISE) && isLeftDriving)
+            {
+                this.invalidationReasons.add(WRONG_WAY_INVALIDATION);
+            }
+
+            // If there is an invalidation logged, invalidate the roundabout
+            if (!this.invalidationReasons.isEmpty())
+            {
+                throw new CoreException("Error building ComplexRoundabout for Edge {}: {}",
+                        source.getIdentifier(), this.invalidationReasons.get(0));
+            }
+        }
+        catch (final Exception exception)
+        {
+            setInvalidReason(exception.getMessage(), exception);
+        }
+    }
+
+    /**
+     * Function for {@link SimpleEdgeWalker} that gathers connected edges that are part of a
+     * roundabout.
+     *
+     * @return {@link Function} for {@link SimpleEdgeWalker}
+     */
+    private Function<Edge, Stream<Edge>> isRoundaboutEdge()
+    {
+        return edge -> edge.connectedEdges().stream().filter(JunctionTag::isRoundabout);
+    }
+
+    public List<String> getInvalidationReasons()
```

Comment:
Semantically, what's the difference between this function and `getAllInvalidations`? It seems like `getError()` could serve the purpose that `getAllInvalidations` is serving here. Then we wouldn't need to override `getAllInvalidations` and could get rid of `getInvalidationReasons`

Reply:
`getAllInvalidations` is not defined in the abstract parent class, so it has to be overridden. I could use `getAllInvalidations` in place of `getInvalidationReasons`, but that means dealing with the rather awkward ComplexEntityError. Really I just wanted a simple method to get what will become the instructions for the MalformedRoundaboutCheck.

Reply:
Good point re: `getAllInvalidations`. Still seems like we have two functions that, by name, should do the same thing, which I don't love. A lot of other ComplexEntities _are_ implementing `getAllInvalidations` the same way, but it also looks like they only have one possible invalidation so the pattern makes a little more sense. 

This isn't a hill I'm willing to die on but I do think using ComplexEntityError is cleaner

Reply:
I figured out a way to use ComplexEntityError in a way I am happy enough with. Let me know what you think.